Add automated performance measurement workflow

joemcgill opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket: https://core.trac.wordpress.org/ticket/57687

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

mukeshpanchal27 committed:
Add performance workflow

mukeshpanchal27 committed:
Added newline at end of file

mukeshpanchal27 committed:
Address review feedback

mukeshpanchal27 committed:
Update workflow steps

mukeshpanchal27 committed:
Add MU plugin for the Server-Timing API

mukeshpanchal27 committed:
Merge branch 'feature/automated-performance-testing-mvp' into add/server-timing-api-mu-plugin

mukeshpanchal27 committed:
Add Configure environment variables step

mukeshpanchal27 committed:
Address review feedback

felixarntz committed:
Merge pull request #6 from 10up/setup-initial-performance-action

Set up initial GitHub action for performance tests

mukeshpanchal27 committed:
Address review feedback

mukeshpanchal27 committed:
Merge branch 'feature/automated-performance-testing-mvp' into add/server-timing-api-mu-plugin

mukeshpanchal27 committed:
Merge branch 'WordPress:trunk' into feature/automated-performance-testing-mvp

mukeshpanchal27 committed:
Merge branch 'WordPress:trunk' into feature/automated-performance-testing-mvp

mukeshpanchal27 committed:
Update MU plugin path

mukeshpanchal27 committed:
Merge pull request #7 from 10up/add/server-timing-api-mu-plugin

MU plugin for the Server-Timing API

mukeshpanchal27 committed:
Merge branch 'WordPress:trunk' into feature/automated-performance-testing-mvp

mukeshpanchal27 committed:
Initial changes

mukeshpanchal27 committed:
Initial changes

mukeshpanchal27 committed:
Add missing file

mukeshpanchal27 committed:
Merge branch 'feature/automated-performance-testing-mvp' into add/server-timing-metrics-gh-workflow

joemcgill committed:
Reinstate previous performance workflow steps

joemcgill committed:
Update the performance tests

This moves the performance tests that make use of the `test-e2e` command from `wp-scripts` into their own directory and updates the specs to fix some JS errors that were previously being triggered.

mukeshpanchal27 committed:
Merge pull request #21 from 10up/add/server-timing-fixes

Performance workflow fixes

mukeshpanchal27 committed:
Fix error after merge

mukeshpanchal27 committed:
Additional changes

joemcgill committed:
Revert performance workflow changes

joemcgill committed:
ignore teset results files

joemcgill committed:
Remove unneccesary package.json changes

joemcgill committed:
Run performance tests during the workflow

joemcgill committed:
Run workflows from the build directory

joemcgill committed:
Make mu-plugins folder 'build'

mukeshpanchal27 committed:
Add dummy data

mukeshpanchal27 committed:
Update repo slug

mukeshpanchal27 committed:
Print result data and remove old files

mukeshpanchal27 committed:
Update XML path. Thanks @joemcgill

Co-authored-by: Joe McGill <801097+joemcgill@users.noreply.github.com>

mukeshpanchal27 committed:
Update .gitignore

Co-authored-by: Joe McGill <801097+joemcgill@users.noreply.github.com>

mukeshpanchal27 committed:
Address review feedback

mukeshpanchal27 committed:
Address review feedback

mukeshpanchal27 committed:
Remove duplicate file

mukeshpanchal27 committed:
Address review feedback

mukeshpanchal27 committed:
Initial changes

mukeshpanchal27 committed:
Address review feedback

joemcgill committed:
Update metrics shape for the dashboard API

joemcgill committed:
Log performance data to codehealth-test for now

joemcgill committed:
Merge pull request #16 from 10up/add/server-timing-metrics-gh-workflow

Record Server-Timing metrics during GH workflows

mukeshpanchal27 committed:
Address review feedback

mukeshpanchal27 committed:
Merge pull request #30 from 10up/add/log-server-metrics-codehealth

Log server metrics to the codehealth dashboard

mukeshpanchal27 committed:
Merge branch 'WordPress:trunk' into feature/automated-performance-testing-mvp

joemcgill committed:
Add support for base metrics

This updates the performance workflow to record base metrics from the 6.1.1 release on every commit in order to normalize the metrics being recorded against a consistent baseline on each run. We do this to reduce the amount of noise that is introduced in the data by changes in the workers that run these measurements.

After recording the initial metrics for the current commit, the docker environment is updated to version 6.1.1 via a WP CLI command and then the metrics are taken again. This time, by passing an optional `--prefix=base` argument so that the results are saved to files with a `base-` prefix for each test. All of the results are then parsed and sent to the dashboard API as part of the `log-results.js` script.

joemcgill committed:
Fix CLI command to update the base environment.

joemcgill committed:
Log plugin path after installing WP in the environment

joemcgill committed:
Make LOCAL_DIR a string

joemcgill committed:
Change the correct file

mukeshpanchal27 committed:
Update build path

joemcgill committed:
Clean up test code from the workflow

mukeshpanchal27 committed:
Fix lint

mukeshpanchal27 committed:
Address review feedback

mukeshpanchal27 committed:
Update function

mukeshpanchal27 committed:
Print base result in workflow

mukeshpanchal27 committed:
Prepare for Core merge

mukeshpanchal27 committed:
Merge branch 'WordPress:trunk' into feature/automated-performance-testing-mvp

joemcgill committed:
Consolidate logic for logging results

mukeshpanchal27 committed:
Address review feedback

joemcgill committed:
Refactor logging logic for clarity

joemcgill committed:
Fix inline comments

joemcgill committed:
Merge pull request #36 from 10up/add/support-base-metrics

Add support for base metrics

joemcgill committed:
Merge branch 'feature/automated-performance-testing-mvp' into prepare-core-merge

joemcgill committed:
Only publish metrics on pushes to trunk

joemcgill committed:
Merge pull request #41 from 10up/prepare-core-merge

Prepare for Core merge

desrosj commented on .github/workflows/performance.yml:
```diff
@@ -0,0 +1,160 @@
+      - name: Install WordPress
+        run: npm run env:install
+
+      - name: Install wordpress importer plugin and import dummy data
```
```suggestion
      - name: Install WordPress Importer plugin and import mock data
```

desrosj replied:
I've never really cared for using `dummy` to describe test data. There are plenty of more inclusive terms we can use.

joemcgill replied:
Noticed later that the same line needs to be updated in the docs at the top of the workflow. Commenting so we don't forget.

youknowriad commented on tests/performance/config/bootstrap.js:
```diff
@@ -0,0 +1,39 @@
+beforeAll( async () => {
+	enablePageDialogAccept();
+
+	//await setupBrowser();
```
Why is this commented?

joemcgill replied:
Good catch, probably was done by mistake during testing.

mukeshpanchal27 replied:
@youknowriad If we run this function then it gives error.

```
Error: Evaluation failed: DOMException: Failed to read the 'localStorage' property from 'Window': Access is denied for this document.
```

youknowriad replied:
I guess clearing local storage shouldn't be done in the initial setupBrowser (because no page is loaded), so maybe we can call directly `await setBrowserViewport( 'large' );`

mukeshpanchal27 replied:
Thanks @youknowriad, I will update it.

youknowriad reviewed (commented):
Very happy to see this 👍

felixarntz commented on .github/workflows/performance.yml:
```diff
@@ -0,0 +1,160 @@
+  push:
+    branches:
+      - trunk
+      - '3.[7-9]'
+      - '[4-9].[0-9]'
+    tags:
+      - '[0-9]+.[0-9]'
+      - '[0-9]+.[0-9].[0-9]+'
+  pull_request:
+    branches:
+      - trunk
+      - '3.[7-9]'
+      - '[4-9].[0-9]'
```
Why do we run this workflow in point release branches, when (at least for this MVP) we only deploy results for `trunk`?

Unless we log the performance results for the other branches _somewhere_, that seems like unnecessary overhead to me. Maybe I'm missing something? Otherwise I would suggest to keep only `trunk` in here.

joemcgill replied:
Even though we're only recording commits from `trunk` to the dashboard for now, I think there's value in having these measurements run all release branches, so we can inspect the action logs when suspect a performance regression has taken place in a point release. Given that commits to release branches are not as frequent, and this workflow is much faster than many others that run on each commit, I don't anticipate this adding a lot of overhead.

Long term, it would be nice to be able to visualize specific releases in dashboard as well, but that is a current limitation of the way the dashboard visualizes commits.

desrosj replied:
Good call out.

Workflows will only run when they exist within the context of the branch being updated. So unless this workflow is backported to WP <= `6.0`, it would not run anyway.

But, I do think that this is a bit too broad. Since there are no plans to backport this branch to older versions, I think we should change this to be "from this point forward".  Similar to what we do in [newer workflows like the PHP Compatibility one](https://github.com/WordPress/wordpress-develop/blob/trunk/.github/workflows/php-compatibility.yml#L5-L19) like this:

```
  push:
    branches:
      - trunk
      - '6.[1-9]'
      - '[7-9].[0-9]'
    tags:
      - '[6-9]+.[0-9]'
      - '[0-9]+.[0-9].[0-9]+'
  pull_request:
    branches:
      - '[0-9]+.[0-9]'
      - '[0-9]+.[0-9].[0-9]+'
      - '![0-5].[0-9].[0-9]+'
      - '!6.0.[0-9]+'
```

felixarntz replied:
@joemcgill Okay, if we are logging the results to the action logs, I agree that that is enough of a justification to have the workflow running for those branches too. 👍

desrosj replied:
Good call out.

Workflows will only run when they exist within the context of the branch being updated. So unless this workflow is backported to WP <= `6.0`, it would not run anyway.

But, I do think that this is a bit too broad. Since there are no plans to backport this branch to older versions, I think we should change this to be "from this point forward".  Similar to what we do in [newer workflows like the PHP Compatibility one](https://github.com/WordPress/wordpress-develop/blob/trunk/.github/workflows/php-compatibility.yml#L5-L19) like this:

```
  push:
    branches:
      - trunk
      - '6.[1-9]'
      - '[7-9].[0-9]'
    tags:
      - '[6-9]+.[0-9]'
      - '[0-9]+.[0-9].[0-9]+'
  pull_request:
    branches:
      - '[0-9]+.[0-9]'
      - '[0-9]+.[0-9].[0-9]+'
      - '![0-5].[0-9].[0-9]+'
      - '!6.0.[0-9]+'
```

desrosj replied:
Sorry, I was mid-review and then GitHub wouldn't let me submit this one comment without submitting the whole review. And then it didn't show me the additional comments above.

> I think there's value in having these measurements run all release branches

I think this is a good point, but we should chat about tradeoffs/maintenance burden of this. Once the workflow is backported we need to backport all updates needed to keep it running.

felixarntz replied:
@desrosj We just commented opposing things at the same time 😆 

I think we may want to eventually get performance data for older branches as well; not necessarily for super old branches, but there's arguably value in making sure that e.g. 6.1.2 is not slower than 6.1.1 even once 6.2 is out. So I think this is something worth pursuing in the future.

> Workflows will only run when they exist within the context of the branch being updated. So unless this workflow is backported to WP <= `6.0`, it would not run anyway.

Is that true? @joemcgill and I had a lengthy conversation about that the other day and based on that I would think that this workflow will run for the other branches too simply because it is in the GitHub repo's main branch `trunk`.

felixarntz replied:
> I think this is a good point, but we should chat about tradeoffs/maintenance burden of this. Once the workflow is backported we need to backport all updates needed to keep it running.

That's a good point as well, though it depends on my last question above and how GitHub workflows actually apply :) If having this here was sufficient to run on other branches, then we wouldn't even need to backport it. But if we did, then I would certainly agree to leave the other branches out of it for now, just to start simple and without overhead. There is value in the data for the other branches too, but at such an early stage I would rather avoid the overhead while we are still establishing how well this all works.

joemcgill replied:
@desrosj good point about needing to backport this to old branches. I think your suggestion to keep this to future branches makes a lot of sense. 

@felixarntz IIRC, the conversation we were having about needing a workflow in the main branch was specifically for a workflow that was meant to run on a cron schedule rather than on a branch event. Hopefully that clarifies things a bit.

desrosj replied:
I think our answer is [here](https://docs.github.com/en/actions/using-workflows/triggering-a-workflow#about-workflow-triggers).

> 1. An event occurs on your repository. The event has an associated commit SHA and Git ref.
> 2. GitHub searches the `.github/workflows` directory in your repository for workflow files that are present in the associated commit SHA or Git ref of the event.
> 3. A workflow run is triggered for any workflows that have `on:` values that match the triggering event. Some events also require the workflow file to be present on the default branch of the repository in order to run.

In our scenario, the ref will be the branch. So if the workflow is not present in that branch, it will not run.

When we [first introduced the workflows into Core](https://core.trac.wordpress.org/browser/trunk/.github/workflows/phpunit-tests.yml?rev=49162), we basically allowed `push` events to all branches using `*.*` and to my recollection, the workflows did not run in the older branches because the workflow files were not present. It's possible this has changed, though.

desrosj replied:
> @felixarntz IIRC, the conversation we were having about needing a workflow in the main branch was specifically for a workflow that was meant to run on a cron schedule rather than on a branch event. Hopefully that clarifies things a bit.

Yes this is accurate. `schedule` events _only_ run on the primary branch. We had to create a way around this for testing older branches receiving security updates regularly. To get around this, we have a `schedule` workflow in `trunk` that uses the GitHub API to create `workflow_dispatch` [calls to each branch for each supported workflow](https://github.com/WordPress/wordpress-develop/blob/trunk/.github/workflows/test-old-branches.yml) in that branch.

After `6.2` is branched, it probably makes sense to add this one to this matrix as well so that it runs at least once a month.

felixarntz replied:
@joemcgill Ahhh thanks for clarifying, makes sense.

In that case, I'd indeed say we should rather remove the other branches from this PR for now.

joemcgill replied:
Updated the branches in [6b51184](https://github.com/WordPress/wordpress-develop/pull/4139/commits/6b511844f20fe21712d526558eb8d30b8ff21655) so this only runs on 6.2+ branches, per @desrosj's suggestion.

desrosj replied:
> After `6.2` is branched, it probably makes sense to add this one to this matrix as well so that it runs at least once a month.

I've created #4160 so we don't forget to add this.

desrosj replied:
I think this conversation can probably be marked resolved. There were a few other inline suggestions I added to further tweak the branch and tag filtering that we can address those individually.

felixarntz commented on tests/performance/log-results.js:
```diff
@@ -0,0 +1,102 @@
+ */
+const parseFile = ( fileName ) => (
+	JSON.parse(
+			fs.readFileSync( path.join( __dirname, '/specs/', fileName ), 'utf8' )
```
Unnecessary extra indentation here :)

felixarntz commented on tests/performance/mu-plugins/server-timing.php:
```diff
@@ -0,0 +1,43 @@
+
+				echo $output;
+			},
+			-9999
```
We may want to set this to `PHP_INT_MIN`. Since WordPress core polyfills that constant, we can safely use it here. The `-9999` is a bit of a hack only needed if that constant doesn't exist.

swissspidy replied:
Isn't it the other way around anyway, the outer hook should be called as early as possible (`PHP_INT_MIN`) and the shutdown hook should run as late as possible (`PHP_INT_MAX`) so that the timing is most accurate?

felixarntz replied:
@swissspidy The outer hook should definitely be `PHP_INT_MAX`, since everything that happens in `template_redirect` should be attributed to the first metric still.

For the inner hook, it may be more debatable. I think we should measure as soon as possible, since at that point the full page has been output, and that's what we care about for the `wp-template` metric. I'm not sure there's anything worth measuring that would happen on `shutdown` anyway.

joemcgill replied:
These have been updated to use the relevant PHP constants. I'm going to resolve this, but feel free to reopen if further discussion in necessary.

felixarntz reviewed (commented):
@joemcgill @mukeshpanchal27 Looks great to me so far. A few small notes below.

felixarntz commented on tests/performance/specs/home-block-theme.test.js:
```diff
@@ -0,0 +1,53 @@
+	} );
+
+	it( 'Server Timing Metrics', async () => {
+		let i = 20;
```
Would be nice to centrally control this somewhere. At a minimum, have it defined in one place, as we should handle the number of runs consistently, not per test scenario.

joemcgill replied:
Handling this in [68e7e66](https://github.com/WordPress/wordpress-develop/pull/4139/commits/68e7e668f693df9d7bb0cab202a906acac93eb18)

joemcgill committed:
Update step name

Co-authored-by: Jonathan Desrosiers <desrosj@users.noreply.github.com>

joemcgill commented on .github/workflows/performance.yml:
```diff
@@ -0,0 +1,160 @@
+            CODEHEALTH_PROJECT_TOKEN: ${{ secrets.CODEHEALTH_PROJECT_TOKEN }}
+            HOST_NAME: "codehealth.vercel.app"
+        # The base hash hard coded here is the commit to the WP 6.1.1 tag.
+        run: node ./tests/performance/log-results.js $CODEHEALTH_PROJECT_TOKEN trunk $GITHUB_SHA 031cedc06d612c4a74f7b6d3e77d8145f2954a25 $COMMITTED_AT $HOST_NAME
```
Update these environment values.

```suggestion
            CODEVITALS_PROJECT_TOKEN: ${{ secrets.CODEVITALS_PROJECT_TOKEN }}
            HOST_NAME: "codevitals.run"
        # The base hash hard coded here is the commit to the WP 6.1.1 tag.
        run: node ./tests/performance/log-results.js $CODEVITALS_PROJECT_TOKEN trunk $GITHUB_SHA 031cedc06d612c4a74f7b6d3e77d8145f2954a25 $COMMITTED_AT $HOST_NAME
```

desrosj replied:
I wanted to double check that this update will keep the secret secure. Looking [through the related documentation](https://docs.github.com/en/actions/security-guides/encrypted-secrets#example-using-bash), it seems that we may want to enclose `$CODEVITALS_PROJECT_TOKEN` in double quotes (`"`).

joemcgill replied:
I did double check this during out internal testing, and the secrets value is not being output anywhere in the logs that I can see.

desrosj replied:
The purpose of the quotes, I believe, is not to obfuscate the secret but rather to prevent any potential issues with special characters. Which doesn't really affect us currently (personally knowing the value of the secret), but could potentially in the future if it should change.

joemcgill replied:
That's helpful. I don't see this token changing often, or at least without our knowledge since it's specifically used for this purpose, so I'm going to resolve this as ok for now.

joemcgill committed:
Update dashboard env variables

joemcgill committed:
Only run the workflow on WP 6.2+ branches

desrosj commented on .github/workflows/performance.yml:
```diff
@@ -0,0 +1,160 @@
+  # - Update permalink structure.
+  # - Run the Performance tests (current commit).
+  # - Print Performance tests results.
+  # - Update the environment to 6.1.1
+  # - Run Performance tests (6.1.1)
+  # - Print Performance tests results (6.1.1).
```
We should remove references to specific versions here to eliminate the need to constantly update these inline docs. Commenting inline below within these steps with ideas how to remove the hardcoded version.

joemcgill replied:
See [my comment below](https://github.com/WordPress/wordpress-develop/pull/4139#discussion_r1122553204) about how the baseline version is meant to be used. Would be happy to update this to something more generic like: "Run Performance tests on baseline version" but I think it's more clear what we're testing if it's explicitly referenced.

desrosj replied:
I recommended extracting the version into a workflow level environment variable below. Maybe just change these comments to remove the actual number reference the baseline version instead.

felixarntz replied:
I think we should do our best to clarify that the idea is **not** that this should always be the latest stable core release. This should be hard-coded to a specific version, which based on where we are now it makes sense to use the current stable core release for. But we shouldn't update it in the future without careful consideration.

I'm sure at some point there will be some good rationale to update it who knows why. But we should do that as rarely as possible to keep the base metric consistent.

joemcgill replied:
I'll iterate on the exact language we use in the inline comment, but am feeling like this is essentially resolved.

desrosj commented on .github/workflows/performance.yml:
```diff
@@ -0,0 +1,160 @@
+
+      - name: Update the environment to 6.1.1
+        run: |
+          npm run env:cli -- core update --version=6.1.1 --force --path=/var/www/build
```
To expand on the comment above for the inline docs, I think it's preferable to remove the hard coded versions in these three steps.

Based on the [WP CLI docs for this command](https://developer.wordpress.org/cli/commands/core/update/), it seems that `core update --force` would be enough to download the latest released version and would eliminate the need for a `--version` flag.

If we really want to display the actual version in these steps, could we add a step earlier that retrieves the latest version dynamically using `https://api.wordpress.org/core/version-check/1.7/` and return it as an output.

desrosj replied:
Thinking about this a bit more. If this runs in older branches, it wouldn't be as simple as grabbing the most recent point release. Say 6.3 is out and the workflow runs in the 6.2 branch. We wouldn't want to compare the performance difference to the latest `6.3.z` release.

I think I'd be OK leaving these 6.1.1 hard coded values for now in the interest of getting this up and running. As long as we circle back to it in the next few weeks and address this.

joemcgill replied:
The purpose that 6.1.1 serves in this workflow is to have a consistent version that serves as a baseline when running performance metrics on the current commit. Including a consistent baseline along with the commit that is being tested allows us to normalize variations in measurements caused by the GitHub runners. In the performance workflow in the Gutenberg repo, this gets updated only once for every major release. However, we're proposing that we leave this pinned to a consistent version, so our baseline stays constant over time, unless we decide to explicitly bump the baseline version.

mukeshpanchal27 replied:
Thanks @desrosj for feedback. If we remove the `--version=6.1.1` then it will install latest version. For now it install `6.2-beta4-55454`

joemcgill replied:
Since the purpose of this step is to install the baseline version we're using, I think it's appropriate to leave it as is for now until we resolve whether want to control this value from an environment variable of some sort.

felixarntz replied:
Correct, we do want to have a hard-coded version here, it must not be always the current version. If we update this every time that a new WordPress release is out, it will skew all future numbers compared to the previous numbers. We can consider changing that in the future, but at least conceptually this needs to be a fixed version.

If we ever change what commit hash this is based on, we will have to acknowledge that performance data after this will not be 100% reliably comparable with the data that was recorded before that change.

desrosj replied:
> Thanks @desrosj for feedback. If we remove the `--version=6.1.1` then it will install latest version. For now it install `6.2-beta4-55454`

Thanks for clarifying that @mukeshpanchal27! I just tested locally as well and I can confirm that is the case. When using the `update` command it will determine whether the site is currently running a development version or not before requesting the update. I tested `wp core download --force`, and that does seem to download the latest version regardless of whether the site is currently using a development version such as `trunk`/beta/rc. But, given the related points around _why_ we want to hard code `6.1.1` to have a consistent base, I think this suggestion can be disregarded.

> Since the purpose of this step is to install the baseline version we're using, I think it's appropriate to leave it as is for now until we resolve whether want to control this value from an environment variable of some sort.

That works for me. I didn't review any of the methodologies behind the metrics being used or reasoning behind what is being reported, just the Actions related code. So fine with leaving the version as 6.1.1 for now.

However, I would like to see this extracted to a workflow level environment variable up with `LOCAL_DIR` This way when we do decide to change the version being used as a baseline, it only requires the version to be changed in one location.

I think it would also be good to include a multi-line doc block above that variable definition explaining why that version is being used for future contributors/us. And there may also be value in adding a docblock at the very top above the workflow name briefly outlining the thinking behind the overall approach being used and why/what the benefit is.

desrosj commented on .github/workflows/performance.yml:
```diff
@@ -0,0 +1,160 @@
+        run: node ./tests/performance/log-results.js $CODEVITALS_PROJECT_TOKEN trunk $GITHUB_SHA 031cedc06d612c4a74f7b6d3e77d8145f2954a25 $COMMITTED_AT $HOST_NAME
+
+      - name: Ensure version-controlled files are not modified or deleted
+        run: git diff --exit-code
```
Can we add the two jobs at the bottom that are in the other Core workflows that run on `push` events for Slack notifications and dispatching a run of the failed workflow... workflow (should probably rename that) for consistency? For example, this is at the [bottom of the PHP compatibility one](https://github.com/WordPress/wordpress-develop/blob/trunk/.github/workflows/php-compatibility.yml#L108).

joemcgill replied:
These have been added, though I'm not sure if we need Slack notifications for this workflow unless we are just looking to be notified about whether the workflow has broken because of an API error. We're definitely not trying to "fail" builds based on these metrics in any way.

desrosj replied:
The notification will post in the `#core` channel of Slack when the workflow encounters an error. Whether that's a network time out, API error, breaking change somewhere in the code base, or GHA service level issue. For the most part I expect these to just run without issue.

In addition to failures being reported in `#core`, a message is posted in `#core-ci-firehose` and `#core-firehose` every time a workflow completes, regardless of the outcome.

We don't currently aggregate workflow outcomes into a authoritative "status badge" like Travis CI had (though it's on a list of things we could explore) because GitHub Actions has one badge per workflow.

Does this seem inline with what you were thinking was alright @joemcgill?

joemcgill replied:
Yeah, sounds good to me. Just wanted to make sure we weren't adding extra noise.

desrosj commented on .github/workflows/performance.yml:
```diff
@@ -0,0 +1,160 @@
+  push:
+    branches:
+      - trunk
+      - '6.[1-9]'
```
Actually, I think the example I gave in the previous discussion was incorrect. If we are not backporting this to <= 6.1 then we should make this exclude 6.1 as well.
```suggestion
      - '6.[2-9]'
```

desrosj commented on .github/workflows/performance.yml:
```diff
@@ -0,0 +1,160 @@
+      - '6.[1-9]'
+      - '[7-9].[0-9]'
+    tags:
+      - '[6-9]+.[0-9]'
```
Wanted to note that this would trigger a build for `6.0` and `6.1` tags, but since these have already been pushed, those won't happen again. We could add an `- '!6.[01]'` exclusion, to better match the branch patterns, but don't feel strongly. I defer to what others feel is appropriate.

joemcgill replied:
Good catch. This has been resolved, I believe.

desrosj commented on .github/workflows/performance.yml:
```diff
@@ -0,0 +1,160 @@
+      - '[7-9].[0-9]'
+    tags:
+      - '[6-9]+.[0-9]'
+      - '[0-9]+.[0-9].[0-9]+'
```
```suggestion
      - '[0-9]+.[0-9].[0-9]+'
      - '![45].[0-9].[0-9]+'
      - '!6.[01].[0-9]+'
```

desrosj replied:
As is, this will match all tags with the X.Y.Z format. This suggestion refines it to exclude all `(4|5).y.z` tags, and `6.0.z`/`6.1.z` tags.

joemcgill replied:
Added in 3a56b3a.

desrosj commented on .github/workflows/performance.yml:
```diff
@@ -0,0 +1,160 @@
+  pull_request:
+    branches:
+      - trunk
+      - '6.[1-9]'
```
```suggestion
      - '6.[2-9]'
```

desrosj commented on .github/workflows/performance.yml:
```diff
@@ -0,0 +1,160 @@
+
+      - name: Install WordPress Importer plugin and import mock data
+        run: |
+          npm run env:cli -- plugin install wordpress-importer --activate
```
The `wordpress-importer` is [already installed as part of the install script](https://github.com/WordPress/wordpress-develop/blob/trunk/tools/local-env/scripts/install.js#L23). So I think we just need to activate it here?

Also, the other WP CLI commands needed `--path=/var/www/build` to run without issue. Do we need the same for this script? I'm curious why only _some_ of the commands need it.
```suggestion
          npm run env:cli -- plugin activate wordpress-importer
```

mukeshpanchal27 replied:
Thanks for the review. If we remove the installation from command then it show `The 'wordpress-importer' plugin could not be found.` error in GH workflow.

Will add `--path=/var/www/build` for scripts. cc. @joemcgill

desrosj replied:
Sorry, I was mistaken in this. The install script does install the plugin, but not in the normal `wp-content` folder. Instead it goes to the `tests/phpunit/data/plugins/wordpress-importer` directory, which makes sense why it wouldn't be present for activation.

We could leave this as is, or we could instead run `cp tests/phpunit/data/plugins/wordpress-importer build/wp-content/plugins` to avoid having to deal with network connections in this step.

desrosj commented on .github/workflows/performance.yml:
```diff
@@ -0,0 +1,160 @@
+
+      - name: Update permalink structure
+        run: |
+          npm run env:cli -- rewrite structure '/%year%/%monthnum%/%postname%/'
```
Same question here about needing `--path`

desrosj commented on .github/workflows/performance.yml:
```diff
@@ -0,0 +1,160 @@
+      - name: Update the environment to 6.1.1
+        run: |
+          npm run env:cli -- core update --version=6.1.1 --force --path=/var/www/build
+          npm run env:cli -- core version --path=/var/www/build
```
We repeat this `--path=/var/www/build` value a handful of times.

Thinking it would be good to set this as an environment variable above with `LOCAL_DIR` and just use something like `${{ env.INSTALL_PATH }}` (don't love that name) throughout?

mukeshpanchal27 replied:
Thanks Jonathan, nice catch. I will use an environment variable for it.

desrosj commented on .github/workflows/performance.yml:
```diff
@@ -0,0 +1,160 @@
+      - name: Install WordPress
+        run: npm run env:install
+
+      - name: Install WordPress Importer plugin and import mock data
```
I think this is fine to leave as is. I personally try to make steps as atomic as possible and would prefer to have them split into two steps: one for activating the importer plugin, one for downloading and importing the mock data (that one could even be split into two but makes sense to leave as one).

If we do split this into two, we need to also chop "and import mock data" off this suggestion.
```suggestion
      - name: Activate the WordPress Importer plugin and import mock data```

desrosj commented on .github/workflows/performance.yml:
```diff
@@ -0,0 +1,160 @@
+        uses: actions/github-script@98814c53be79b1d30f795b907e553d8679345975 # v6.4.0
+        id: commit-timestamp
+        with:
+            github-token: ${{secrets.GITHUB_TOKEN}}
```
```suggestion
            github-token: ${{ secrets.GITHUB_TOKEN }}
```

Super small nit to add spaces. Just matches the syntax used in the other expressions here and within other workflows.

desrosj commented on .github/workflows/performance.yml:
```diff
@@ -0,0 +1,160 @@
+            github-token: ${{secrets.GITHUB_TOKEN}}
+            script: |
+                const commit_details = await github.rest.git.getCommit({owner: context.repo.owner, repo: context.repo.repo, commit_sha: context.sha});
+                return parseInt((new Date( commit_details.data.author.date ).getTime() / 1000).toFixed(0))
```
At first I wondered if we should check for an error to be returned by the REST API, but I think an error happening in either line is fine. We just need a failure to occur.

joemcgill replied:
I "tested" this and if the API requests encounters an error, the whole step (and thus the workflow) will fail, which I think is what we want. However, these only need to be set when we're publishing results, so we can add the same `if` value that we're using for that step to keep these from running when they're not being used.

joemcgill replied:
Added conditionals. Let me know if you think we should add more error handling.

desrosj replied:
I think this works. 👍🏼

desrosj commented on tests/performance/config/bootstrap.js:
```diff
@@ -0,0 +1,39 @@
+/**
+ * WordPress dependencies
```
Should this have a `.`?

joemcgill replied:
Fixed in dca77c1

desrosj commented on .github/workflows/performance.yml:
```diff
@@ -0,0 +1,160 @@
+  # - Logs Docker debug info.
+  # - Installs WordPress.
+  # - Installs the Server Timing MU plugin.
+  # - Install wordpress importer plugin and import dummy data.
```
After we're happy with any changes below we should do a full step through of these docs just to confirm they're still accurate.

joemcgill replied:
I've updated this list to match the names of each step for consistency in f5446a5

desrosj commented on .github/workflows/performance.yml:
```diff
@@ -0,0 +1,160 @@
+      - name: Install MU plugin
+        run: |
+          mkdir ./build/wp-content/mu-plugins
+          cp ./tests/performance/mu-plugins/server-timing.php ./build/wp-content/mu-plugins/server-timing.php
```
I'm not familiar with best practices for Jest testing setups. But my first thought was "should this `mu-plugins` folder be moved within a `data` folder similar to how our PHPUnit folder is structured to make it more clear that the `mu-plugins` directory is meant to be copied into the install before running tests?"

joemcgill replied:
I don't have a strong preference on naming conventions here. This folder/file only exists to be copied into the environment during the workflow. If it's clearer to move this into a `data` folder, I'm happy to do so.

desrosj replied:
To elaborate on my thinking a bit. I was thinking what if a plugin or theme is added in the future. Then we'd have a `themes` and `plugins` directory in this main `tests/performance` folder as well. We could always make a decision if that should happen.

@felixarntz or @youknowriad any preference here? Don't feel strongly either way.

felixarntz replied:
How about we move this to `tests/performance/wp-content/mu-plugins`? `wp-content` as a containing folder would be a very clear name for everyone I think :)

desrosj replied:
> How about we move this to `tests/performance/wp-content/mu-plugins`? `wp-content` as a containing folder would be a very clear name for everyone I think :)

Works for me!

joemcgill replied:
Handled in [62a288d](https://github.com/WordPress/wordpress-develop/pull/4139/commits/62a288d2c8b2222dfc3f471404de1c1cd2ed42b4)

desrosj commented on .github/workflows/performance.yml:
```diff
@@ -0,0 +1,160 @@
+            CODEVITALS_PROJECT_TOKEN: ${{ secrets.CODEVITALS_PROJECT_TOKEN }}
+            HOST_NAME: "codevitals.run"
+        # The base hash hard coded here is the commit to the WP 6.1.1 tag.
+        run: node ./tests/performance/log-results.js $CODEVITALS_PROJECT_TOKEN trunk $GITHUB_SHA 031cedc06d612c4a74f7b6d3e77d8145f2954a25 $COMMITTED_AT $HOST_NAME
```
If we make the workflow more dynamic in the other areas, would we be able to pass a branch or tag name instead of a SHA value here? If not, then this SHA would also need to be fetched through a REST API call, or something similar.

joemcgill replied:
This value is used for logging purposes, but if we do want to make the baseline version variable, we could use the `gitTag` method on the API Client to get this value based on a tag reference.

See: https://octokit.github.io/rest.js/v19#git-get-tag

joemcgill replied:
Handled in [04c0e39](https://github.com/WordPress/wordpress-develop/pull/4139/commits/04c0e395fd1c6e72a29fd3809d9c0b9b6943b3f0)

desrosj commented on .github/workflows/performance.yml:
```diff
@@ -0,0 +1,160 @@
+  # - Publish performance results.
+  # - Ensures version-controlled files are not modified or deleted.
+  performance:
+    runs-on: ubuntu-latest
```
Can we add a `name:` here? The `name:` shows up in the UI. While the default of the job name `performance` is fine, how about `name: Configure & run tests`, or something to that effect describing what happens in the job.

joemcgill committed:
Update workflow branch to 6.2+

Co-authored-by: Jonathan Desrosiers <desrosj@users.noreply.github.com>

mukeshpanchal27 committed:
Address review feedback for bootstrap.js

mukeshpanchal27 committed:
add feature branch

mukeshpanchal27 committed:
Remove setupBrowser() from beforeAll function

mukeshpanchal27 committed:
Address feedback

mukeshpanchal27 committed:
Remove static WP version

mukeshpanchal27 committed:
Revert "Remove static WP version"

This reverts commit a065f0a93cbbe081115c9dfdf772eb767d1117e7.

mukeshpanchal27 committed:
Add Slack Notifications

mukeshpanchal27 committed:
Correct tags

mukeshpanchal27 committed:
Correct tags

mukeshpanchal27 committed:
Active plugin

mukeshpanchal27 committed:
Add path

mukeshpanchal27 committed:
Add path

mukeshpanchal27 committed:
Add path

mukeshpanchal27 committed:
Use variable for build path

mukeshpanchal27 committed:
Separate workflow steps

mukeshpanchal27 committed:
Add period at the end

mukeshpanchal27 committed:
Add job name

mukeshpanchal27 committed:
Remove feature branch

mukeshpanchal27 committed:
Update bootstrap.js

mukeshpanchal27 committed:
Add feature branch

joemcgill committed:
Simplify tags matrix

joemcgill committed:
Remove feature branches from matrix

This was used for testing changes before merging to the feature branch.

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

joemcgill committed:
Merge pull request #43 from 10up/fix/address-review-feedback

Address review feedback for core PR

joemcgill committed:
Inline docs fix

desrosj commented on .github/workflows/performance.yml:
```diff
@@ -0,0 +1,208 @@
+      - '[7-9].[0-9]'
+    tags:
+      - '[6-9]+.[2-9]'
+      - '[6-9]+.[2-9].[0-9]+'
```
I think our branch filters are now correct. But a few more tweaks on the tag filtering.

This will match all version tags 6.2-6.9, but then only matches 7.2-7.9, 8.2-8.9, etc.

Let's just go with this:

```
      - '[0-9]+.[0-9]'
      - '[0-9]+.[0-9].[0-9]+'
      - '![45].[0-9].[0-9]+'
      - '!6.[01].[0-9]+'

```

This will:
- include all `X.Y` version tags (we know there will never be another `6.0`, `4.5`, etc.)
- include all `X.Y.Z` version tags
- Selectively exclude all `X.Y.Z` versions that start with `4.`, or `5.` while allowing `45.Y.Z`, etc. versions to work (that would be a wild version to reach, but if we are ever able to switch to more rapid releases, it could happen)
- Selectively exclude all `X.Y.Z` versions that start with `6.0.` or `6.1.` while allowing `60.Y.Z`

joemcgill replied:
Of cousre! I think @mukeshpanchal27 had this right and I got too cute trying to simplify it 🙃

joemcgill replied:
This is now fixed 🤞🏻

desrosj commented on .github/workflows/performance.yml:
```diff
@@ -0,0 +1,208 @@
+
+env:
+  LOCAL_DIR: build
+  BUILD_PATH: '--path=/var/www/build'
```
Don't feel strongly about this, but when I suggested this environment variable, I envisioned keeping `--path=` inline for the CLI command, and using it like `--path=${{ env.BUILD_PATH }}`.

Don't think we need the raw path anywhere right now, but could come in handy to have in the future?

joemcgill replied:
I wondered the same thing, or even remove the extra environment var and do `--path=/var/www/${{ env.LOCAL_DIR }}` since those two should always be in sync.

joemcgill replied:
Updated in [2e15306](https://github.com/WordPress/wordpress-develop/pull/4139/commits/2e153061a54d7fbbb46cf7b356a18285d61f2931)

mukeshpanchal27 replied:
Thanks Joe.  I misunderstood it.

joemcgill committed:
Set the base version via env variable

This adds a new BASE_HASH environment variable that is used throughout the workflow where we reference a base version or hash.

felixarntz commented on .github/workflows/performance.yml:
```diff
@@ -0,0 +1,224 @@
+
+      - name: Install MU plugin
+        run: |
+          mkdir ./build/wp-content/mu-plugins
+          cp ./tests/performance/mu-plugins/server-timing.php ./build/wp-content/mu-plugins/server-timing.php
```
We should change the hard-coded `build` directory to reference the `LOCAL_DIR` env variable instead.

joemcgill replied:
Also handled in [2e15306](https://github.com/WordPress/wordpress-develop/pull/4139/commits/2e153061a54d7fbbb46cf7b356a18285d61f2931)

felixarntz commented on .github/workflows/performance.yml:
```diff
@@ -23,8 +23,14 @@ concurrency:
+  # and another against a consistent version that is used as a baseline measurement.
+  # This is done to isolate variance in measurements caused by the GitHub runners
+  # from differences caused by code changes between commits. The BASE_TAG value here
+  # represents the current version being used for baseline measurements.
```
See my comment in https://github.com/WordPress/wordpress-develop/pull/4139#discussion_r1123703622, I think this comment is great, but I think we should further clarify that this should not be updated to always be the current WordPress version.

I think the comment as is looks _almost_ good in that sense, just that last bit on "represents the current version being used" may be misunderstood as if it should be the current WordPress version. Maybe we can clarify that further?

joemcgill replied:
@felixarntz Do you think this is more clear?

```
  # This workflow takes two sets of measurements — one for the current commit,
  # and another against a consistent version that is used as a baseline measurement.
  # This is done to isolate variance in measurements caused by the GitHub runners
  # from differences caused by code changes between commits. The BASE_TAG value here
  # represents the version being used for baseline measurements. It should only be
  # changed if we want to normalize results against a different baseline.
```

felixarntz replied:
@joemcgill Thanks, that sounds a lot better to me 👍

joemcgill replied:
Updated in ccd09a0

joemcgill committed:
Use LOCAL_DIR for all paths and remove BUILD_PATH

joemcgill committed:
Move mu-plugins test directory

joemcgill committed:
Just another tags matrix update

joemcgill committed:
Only get repo info when needed for publishing

joemcgill committed:
Fix path to test mu-plugin dir

joemcgill committed:
Configure the number of test runs in one place

costdev commented on tests/performance/config/bootstrap.js:
```diff
@@ -0,0 +1,39 @@
+	await setBrowserViewport( 'large' );
+}
+
+// Before every test suite run, delete all content created by the test. This ensures
+// other posts/comments/etc. aren't dirtying tests and tests don't depend on
+// each other's side-effects.
```
```suggestion
/*
 * Before every test suite run, delete all content created by the test. This ensures
 * other posts/comments/etc. aren't dirtying tests and tests don't depend on
 * each other's side-effects.
 */
```
Just a very minor multi-line comment format change to meet the JS documentation standards.

costdev reviewed (approved):
Thanks for the ping to review this PR @mukeshpanchal27!

Aside from a minor comment format change, I didn't find any areas of concern that weren't already mentioned and resolved in earlier discussions. Nice work all!

mukeshpanchal27 committed:
Update comment format

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

joemcgill committed:
Update inline comment about the BASE_TAG

joemcgill committed:
Update table of contents to match step names

felixarntz reviewed (approved):
@joemcgill @mukeshpanchal27 Awesome work, LGTM! 🚢 

This is a great starting point to go from. Once this is merged, let's assess the variance of the metrics we get to see how reliable they are. There are several follow up tickets to open for further exploration, e.g. additional metrics (most importantly something for the client-side like LCP, and maybe Web Vitals in general) or additional scenarios (potentially a translated version of WP, a site with an object cache, ideas like that). Of course we also have to keep in mind to keep the duration of the workflow manageable, so those decisions will need to be taken with care.

desrosj commented on .github/workflows/performance.yml:
```diff
@@ -0,0 +1,231 @@
+      - name: Print base line performance tests results
+        run: "node ./tests/performance/results.js --prefix=base"
+
+      - name: Set the base sha
```
```suggestion
      - name: Set the base SHA
```

Acronym.

desrosj reviewed (approved):
I'm happy with all the revisions. Thanks so much all! Really looking forward to getting this merged so we can start tracking performance changes changeset to changeset.

joemcgill commented:
Committed in [55459](https://core.trac.wordpress.org/changeset/55459)
